crates/driver/src/domain/competition/solution/fee.rs

```diff
@@ -78,21 +78,54 @@ impl Fulfillment {
             Some(FeePolicy::Surplus {
                 factor,
                 max_volume_factor,
-            }) => {
-                let fee_from_surplus = self.fee_from_surplus(prices, *factor)?;
-                let fee_from_volume = self.fee_from_volume(prices, *max_volume_factor)?;
-                // take the smaller of the two
-                tracing::debug!(uid=?self.order().uid, fee_from_surplus=?fee_from_surplus, fee_from_volume=?fee_from_volume, protocol_fee=?(std::cmp::min(fee_from_surplus, fee_from_volume)), executed=?self.executed(), surplus_fee=?self.surplus_fee(), "calculated protocol fee");
-                Ok(std::cmp::min(fee_from_surplus, fee_from_volume))
-            }
+            }) => self.calculate_fee(
+                self.order().sell.amount.0,
+                self.order().buy.amount.0,
+                prices,
+                *factor,
+                *max_volume_factor,
+            ),
+            Some(FeePolicy::PriceImprovement {
+                factor,
+                max_volume_factor,
+                quote,
+            }) => self.calculate_fee(
+                quote.sell_amount,
+                quote.buy_amount,
```

Comment:
I think here we need to use whatever is better (quote or limit price)

Reply:
https://github.com/cowprotocol/services/pull/2277/commits/f629e6138270ce124f3cbd043d3d2008012ec4ca